shakedown/dcos/command.py

```diff
@@ -158,3 +172,78 @@ def run_dcos_command(command, raise_on_error=False, print_output=True):
             return_code, command, stdout, stderr))
 
     return stdout, stderr, return_code
+
+
+class HostSession:
+    """Context manager that returns an SSH session to run commands.
+    
+    """
+    def __init__(self, host, username, key_path, verbose):
+        self.host = host
+        self.username = username
+        self.key_path = key_path
+        self.verbose = verbose
+        self.exit_code = -1
+        self.output = ''
+        self.session = None
+    
+    def __enter__(self):
+        """Called when using a `with` closure.
+
+        :return: this session manager
+        :rtype: HostSession
+        """
+        self.session = get_session(
+                self.host,
+                self.username,
+                self.key_path)
+        return self
+    
+    def __exit__(self, *args):
+        """Executed when the context manager is complete (exits
+        the with closure).
+
+        :return: None
+        """
+        self.exit_code = self.session.recv_exit_status()
+        self._wait_for_recv()
+        # read data that is ready
+        while self.session.recv_ready():
+            # lists of file descriptors that are ready for IO
+            # read, write, "exceptional condition" (?)
+            rl, wl, xl = select([self.session], [], [], 0.0)
+            if len(rl) > 0:
+                recv = str(self.session.recv(1024), "utf-8")
+                if self.verbose:
+                    print(recv, end='', flush=True)
+                self.output += recv
+        try_close(self.session)
+        return None
+    
+    def _wait_for_recv(self):
+        """After executing a command, wait for results.
+        
+        Because `recv_ready()` can return False, but still have a
+        valid, open connection, it is not enough to ensure output
+        from a command execution is properly captured.
+
+        :return: None
+        """
+        while True:
+            time.sleep(0.2)
+            if self.session.recv_ready() or self.session.closed:
+                return
+    
+    def run(self, command):
+        """Run `command` on this SSH session. This does not return the
+        result, use `get_result` to retrieve command's results.
+
+        :param command: SSH command to run
+        :type command: str
+        
+        :return: None
+        """
+        self.session.exec_command(command)
```

Comment:
If you put all the code from `__exit__` here, you could run multiple commands on a shared ssh session. Maybe it's not important, but feels non-intuitive to only be able to use it once.

```
with HostSession(host, username, key_path, noisy) as s:
    ec1, output1 = s.run(cmd1)
    ec2, output2 = s.run(cmd2)
```

Reply:
The `run_command` method only accepts one command at a time. It is used in tests as follows:

```
success, output = run_command(master(), 'ps aux', ....)
if output.contains(....):
  # do something
success, output = run_command(master(), 'iptables -L', ...)
...
```